pkg/codec/vaapi/vp8.go

```diff
@@ -0,0 +1,491 @@
+	"github.com/pion/mediadevices/pkg/prop"
+
+	"github.com/pion/webrtc/v2"
+)
+
+const (
+	surfaceVP8Ref0 = iota
+	surfaceVP8Ref1
+	surfaceVP8Ref2
+	surfaceVP8Ref3
+	surfaceVP8Input
+	surfaceVP8Num
+)
+
+type encoderVP8 struct {
+	r     video.Reader
+	buf   []byte
+	frame []byte
+
+	fdDRI    C.int
+	display  C.VADisplay
+	confID   C.VAConfigID
+	surfs    [surfaceVP8Num]C.VASurfaceID
+	ctxID    C.VAContextID
+	seqParam C.VAEncSequenceParameterBufferVP8
+	picParam C.VAEncPictureParameterBufferVP8
+	qMat     C.VAQMatrixBufferVP8
+	hrdParam hrdParam
+	frParam  frParam
+	rcParam  rcParam
+
+	frameCnt int
+	prop     prop.Media
+
+	mu     sync.Mutex
+	closed bool
+}
+
+func init() {
+	codec.Register(webrtc.VP8, codec.VideoEncoderBuilder(NewVP8Encoder))
+}
+
+// NewVP8Encoder creates new VP8 encoder
+func NewVP8Encoder(r video.Reader, p prop.Media) (io.ReadCloser, error) {
+	if (p.Width/16)*16 != p.Width || p.Width == 0 {
+		return nil, errors.New("width must be 16*n")
+	}
+	if (p.Height/16)*16 != p.Height || p.Height == 0 {
+		return nil, errors.New("height must be 16*n")
+	}
+	if p.KeyFrameInterval == 0 {
+		p.KeyFrameInterval = 30
+	}
+
+	// Parameters are from libva-utils/encode/vp8enc.c
+	e := &encoderVP8{
+		r:    video.ToI420(r),
+		prop: p,
+		seqParam: C.VAEncSequenceParameterBufferVP8{
+			frame_width:     C.uint(p.Width),
+			frame_height:    C.uint(p.Height),
+			bits_per_second: C.uint(p.BitRate),
+			intra_period:    C.uint(p.KeyFrameInterval),
+			reference_frames: [4]C.VASurfaceID{
+				C.VA_INVALID_ID,
+				C.VA_INVALID_ID,
+				C.VA_INVALID_ID,
+				C.VA_INVALID_ID,
+			},
+		},
+		picParam: C.VAEncPictureParameterBufferVP8{
+			ref_last_frame:      C.VA_INVALID_SURFACE,
+			ref_gf_frame:        C.VA_INVALID_SURFACE,
+			ref_arf_frame:       C.VA_INVALID_SURFACE,
+			reconstructed_frame: C.VA_INVALID_SURFACE,
+			clamp_qindex_low:    9,
+			clamp_qindex_high:   127,
+			loop_filter_level: [4]C.int8_t{
+				19, 19, 19, 19,
+			},
+		},
+		qMat: C.VAQMatrixBufferVP8{
+			quantization_index: [4]C.uint16_t{
+				60, 60, 60, 60,
+			},
+			quantization_index_delta: [5]C.int16_t{
+				0, 0, 0, 0, 0,
+			},
+		},
+		hrdParam: hrdParam{
+			hdr: C.VAEncMiscParameterBuffer{
+				_type: C.VAEncMiscParameterTypeHRD,
+			},
+			data: C.VAEncMiscParameterHRD{
+				initial_buffer_fullness: C.uint(p.BitRate * 1500 / 2),
+				buffer_size:             C.uint(p.BitRate * 1500),
+			},
+		},
+		frParam: frParam{
+			hdr: C.VAEncMiscParameterBuffer{
+				_type: C.VAEncMiscParameterTypeFrameRate,
+			},
+			data: C.VAEncMiscParameterFrameRate{
+				framerate: C.uint(p.FrameRate),
+			},
+		},
+		rcParam: rcParam{
+			hdr: C.VAEncMiscParameterBuffer{
+				_type: C.VAEncMiscParameterTypeRateControl,
+			},
+			data: C.VAEncMiscParameterRateControl{
+				window_size:       1500,
+				initial_qp:        60,
+				min_qp:            9,
+				bits_per_second:   C.uint(float32(p.BitRate) * 1.25),
+				target_percentage: C.uint(100.0 / 1.25),
```

Comment:
seems better to have `p.MaxBitRate` prop to get intended balance of quality and bitrate.